lib/ocpp/common/websocket/websocket_tls_tpm.cpp

```diff
@@ -600,7 +620,8 @@ void WebsocketTlsTPM::on_conn_connected() {
     this->m_is_connected = true;
     this->reconnecting = false;
 
-    this->connected_callback(this->connection_options.security_profile);
+    std::thread connected([this]() { this->connected_callback(this->connection_options.security_profile); });
```

Comment:
We need to run it in a different thread on reconnect only. That is actually the bug that was present that lead to loss of the first message after reconnect. 

WebsocketTlsTPM::on_conn_connected() is called from the client_loop thread and calls connected_callback which in turn sends a status notification which then calls send -> poll_message() ->msg_send_cv.wait_for. The cv waits to set msg->message_sent to true on the message but the client_loop is actually waiting for on_conn_connected() to finish first. Once the timeout of the cv is expired it says the message was dropped and immediatly after it is being sent because the client_loop runs again.
After that it works normally.
Whether that can happen on close too? We need to check

Reply:
So i think it is a similar deadlock as calling the other callback_handlers directly from the client_loop

Reply:
In that case we need to make sure that all the callbacks are not sent from the client_loop to prevent the case in which a message is polled (and waited) from the client thread, and all callbacks should be called like this one, on a different thread.

Reply:
this->closed_callback(websocketpp::close::status::normal);

Reply:
Most likely the handler will not send new messages on this callback, but if it does, it deadlocks here too. I'll add a thread around this just to be sure.